File: sol/solution/0001-0100/0022/Solution.ex
defmodule Solution do
  @spec generate_parenthesis(n :: integer) :: [String.t]
  def generate_parenthesis(n) do
    generate([], "", 0, 0, n)
  end
  defp generate(acc, curr, open, close, n) do
    if String.length(curr) == n * 2 do
      [curr | acc]
    else
      acc =
        if open < n do
          generate(acc, curr <> "(", open + 1, close, n)
        else
          acc
        end
      if close < open do
        generate(acc, curr <> ")", open, close + 1, n)
      else
        acc
      end
    end
  end
end

File: sol/solution/0001-0100/0022/Solution.erl
-spec generate_parenthesis(N :: integer()) -> [unicode:unicode_binary()].
generate_parenthesis(N) ->
    generate(N, N, <<>>).
generate(0, 0, Acc) ->
    [Acc];
generate(Open, Close, Acc) when Open > 0, Open =< Close ->
    generate(Open - 1, Close, <<Acc/binary, "(">>) ++
    (if Close > Open -> generate(Open, Close - 1, <<Acc/binary, ")">>); true -> [] end);
generate(Open, Close, Acc) when Close > Open ->
    generate(Open, Close - 1, <<Acc/binary, ")">>);
generate(_, _, _) ->
    [].